### ISSUE TITLE
Transifex appears to require credentials for pulling translations

### ISSUE THREAD

**Issue by ferventcoder** [cake-contrib/Cake.Recipe#296]
Using ChocolateyGUI without having transifex configured, I didn't always see an error. Then a few months ago I started seeing an error and it appears that maybe the side of transifex that did not require credentials now does. I dug into it at https://github.com/cake-contrib/Cake.Recipe/blob/97c342efee19f38a43c3e6aafef18386efdda199/Cake.Recipe/Content/transifex.cake#L18-L55. Not having Transifex environment variables set does the following in a diagnostic log:

~~~
========================================
Transifex-Setup
========================================
Skipping task: Transifex-Setup

========================================
Transifex-Push-SourceFiles
========================================
Skipping task: Transifex-Push-SourceFiles

========================================
Transifex-Pull-Translations
========================================
Executing task: Transifex-Pull-Translations
Resolving assembly Cake.Transifex.resources, Version=0.7.0.0, Culture=en-US, PublicKeyToken=null
Exception while resolving assembly Cake.Transifex.resources: Could not load file or assembly 'Cake.Transifex.resources' or one of its dependencies. The system cannot find the file specified.
Executing: "C:/ProgramData/Chocolatey/bin/tx.exe" pull --all --minimum-perc=60 --mode=onlytranslated
HttpNotAuthorized: Authorization Required
An error occurred when executing task 'Transifex-Pull-Translations'.
~~~

**Comment by ferventcoder** [comment]
Going right to Transifex:

~~~
> & "C:/ProgramData/chocolatey/bin/tx.exe" pull --all --minimum-perc=60 --mode=onlytranslated
HttpNotAuthorized: Authorization Required
~~~

**Comment by ferventcoder** [comment]
Upgraded to latest and this is the error I'm seeing now:

~~~
 C:\ > & "C:/ProgramData/chocolatey/bin/tx.exe" --version
0.13.2, py 3.6, AMD64
C:\ > & "C:/ProgramData/chocolatey/bin/tx.exe" pull --all --minimum-perc=60 --mode=onlytranslated
tx ERROR: Authorization Required
tx ERROR:
Error: Authentication failed. Please make sure your credentials are valid. You
can update your credentials in the ~/.transifexrc file. For more information,
visit https://docs.transifex.com/client/client-configuration#-transifexrc.
~~~

**Comment by pascalberger** [comment]
@AdmiringWorm Do you have an idea what might be the cause of this?

**Comment by ferventcoder** [comment]
https://github.com/transifex/transifex-client/issues/100

**Comment by ferventcoder** [comment]
* https://docs.transifex.com/client/introduction#section-authenticating
* https://docs.transifex.com/client/pull

Not sure when it was changed, but the documentation does say to use the client you need to provide authentication.

**Comment by ferventcoder** [comment]
Copying these lines down as criteria for pulling enables the build to work:

https://github.com/cake-contrib/Cake.Recipe/blob/97c342efee19f38a43c3e6aafef18386efdda199/Cake.Recipe/Content/transifex.cake#L24-L25

to right after https://github.com/cake-contrib/Cake.Recipe/blob/97c342efee19f38a43c3e6aafef18386efdda199/Cake.Recipe/Content/transifex.cake#L46

**Comment by ferventcoder** [comment]
Also note, at some point I created a `~/.transifexrc` file with the following details:

~~~
[https://www.transifex.com]
api_hostname = https://api.transifex.com
hostname = https://www.transifex.com
password = no
username = no
~~~

**Comment by ferventcoder** [comment]
If I remove that file, it will ask me to create it again

~~~
 & "C:/ProgramData/chocolatey/bin/tx.exe" pull --all --minimum-perc=60 --mode=onlytranslated
tx INFO: No credentials file was found at C:\Users\rob\.transifexrc.
tx INFO: Created C:\Users\rob\.transifexrc
[?] Enter your api token:
Verifying token...
tx ERROR: Authorization credentials are missing. Make sure that you have run `tx init` to setup your credentials.
tx ERROR: 'Api' object has no attribute 'username'
~~~

**Comment by ferventcoder** [comment]
Once I add a valid token to that, everything works. I didn't attempt to run the build between renaming the file and rerunning tx directly. It could be that the mere existence of the file enables transifex pulling, which caused my issue. It might be better to capture the error and explain user next steps for determining the issue - actionable errors are really important for folks who are not wanting to dig into the depths to work on ancillary things can just move on to the areas we need to be focused in and want to dig in. If that makes sense.

**Comment by ferventcoder** [comment]
I might also have the file on the system but NOT want it to automatically opt me into transifex pulls

**Comment by gep13** [comment]
Okay, all is right with the world again...

I have had a look at this, and I am able to explain what is going on, and why I was struggling to understand what is going on.  Let me try to explain...

My first thought was that Cake.Recipe is being used on over 100 repos in the Cake-Contrib Organisation, so why is this not causing a problem on those machines.  The answer to that question is here:

https://github.com/cake-contrib/Cake.Recipe/blob/develop/Cake.Recipe/Content/build.cake#L571

The `Pull-Transifex-Translations` task is only added to the DAG when the `TransifexEnabled` property is true.  And this is only set to true if it is forced (via SetParameters) or based on this:

https://github.com/cake-contrib/Cake.Recipe/blob/develop/Cake.Recipe/Content/transifex.cake#L15

i.e. if a `./.tx/config` file exists in the repository.  None of the Cake-Contrib proiects have either of those things, so that explains why this isn't a problem there.  However, ChocolateyGUI does have that file in the repository, and as a result, it has a problem.  So, now onto that...

I have just set up a new VM, that has NOTHING on it, other than Windows Server 2016.  On to it I installed:

* Chocolatey
* Git
* dotnet 4.7.2

I then cloned the ChocolateyGUI repository, and ran the build (`./build.ps1`) and this is what I got:

![image](https://user-images.githubusercontent.com/1271146/49571561-926b0100-f931-11e8-9552-8fb1173e59c4.png)

So, @ferventcoder seems to be correct that pulling translations seems to need authentication, which I don't think was required before.  I can confirm on my machine I have a `.transifexrc` file:

![image](https://user-images.githubusercontent.com/1271146/49571826-3654ac80-f932-11e8-90d4-67aea4d89cfd.png)

with the correct API token in it, and that is why the pull of the translations is working on my machine.

But, then the obvious question is how/why is this working on AppVeyor...

I have just setup an RDP connection to the AppVeyor Build Agent, and I can confirm that at the start of the build, there was no `.transifexrc` file in place, however, as soon as the Transifex tasks started, it was created.

Back on my VM, I ran the build again, which results in the following:

![image](https://user-images.githubusercontent.com/1271146/49572501-dced7d00-f933-11e8-9ac0-5c3593ae190b.png)

Which is what @ferventcoder is reporting.

I then went and deleted the generated `.transifexrc` file, added the environment variable, and ran the build again, and got the following:

![image](https://user-images.githubusercontent.com/1271146/49572796-77e65700-f934-11e8-8a20-acebc40cb3b6.png)

Which I actually expected to work.  Digging further, after looking at what tasks are executed on AppVeyor, brought be to the following:

https://github.com/cake-contrib/Cake.Recipe/blob/develop/Cake.Recipe/Content/transifex.cake#L18-L32

So, long story short, the AppVeyor build, when Transifex is enabled, and has credentials, writes out the necessary `.transifexrc` file to make everything work correctly.

As suggested, we need additional checks on the Pull-Translations task to match what is done with the others, i.e. should take a dependency on the setup task, and only run if credentials are present.

At some point, on my machine, I must have run the setup task, and had forgotten about it, or entered the apitoken when it was requested, which created the file with valid information.  If you input incorrect information when prompted, when running the build a second time, it produces the unauthorized error.

**Comment by gep13** [comment]
Well, that was an interesting way to start the morning, lots of head scratching and debugging there.

**Comment by gep13** [comment]
@ferventcoder I have just tested the change that @AdmiringWorm has suggested in his PR, and happy to say that your problem has now been resolved.
